ext/random/random.c

```diff
@@ -0,0 +1,1763 @@
+		if (str_seed->len == 32) {
+			/* Endianness safe copy */
+			int i;
+			char *ptr = (char *) &state->s;
+			for (i = 0; i < 32; i++) {
+				ptr[i] = str_seed->val[i];
+			}
+		}  else {
+			zend_argument_value_error(1, "state strings must be 32 bytes");
+			RETURN_THROWS();
+		}
+	} else {
+		engine->algo->seed(state, int_seed);
+	}
+}
+/* }}} */
+
+/* {{{ Construct object */
+PHP_METHOD(Random_Randomizer, __construct)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_object *engine_object = 0;
+	zval zengine_object;
+	zend_string *pname;
+	
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(engine_object, random_ce_Random_Engine);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Create default RNG instance */
+	if (!engine_object) {
+		engine_object = php_random_engine_secure_new(random_ce_Random_Engine_Secure);
+		zend_call_known_instance_method_with_0_params(
+			random_ce_Random_Engine_Secure->constructor,
+			engine_object,
+			NULL
+		);
+
+		/* No need self-refcount */
+		GC_DELREF(engine_object);
+	}
+
+	ZVAL_OBJ(&zengine_object, engine_object);
+
+	/* Write property */
+	pname = zend_string_init("engine", sizeof("engine") - 1, 0);
+	zend_std_write_property(&randomizer->std, pname, &zengine_object, NULL);
+	zend_string_release(pname);
+
+	php_random_randomizer_common_init(randomizer, engine_object);
+}
+/* }}} */
+
+/* {{{ Generate random number in range */
+PHP_METHOD(Random_Randomizer, getInt)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_long min, max;
+	int argc = ZEND_NUM_ARGS();
+
+	if (argc == 0) {
+		// right shift for compatibility
+		RETURN_LONG(((uint64_t) randomizer->algo->generate(randomizer->state)) >> 1);
+	}
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_LONG(min)
+		Z_PARAM_LONG(max)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (UNEXPECTED(max < min)) {
+		zend_argument_value_error(2, "must be greater than or equal to argument #1 ($min)");
+		RETURN_THROWS();
+	}
+
+	RETURN_LONG(php_random_engine_range(randomizer->algo, randomizer->state, min, max));
+}
+/* }}} */
+
+/* {{{ Generate random bytes string in ordered length */
+PHP_METHOD(Random_Randomizer, getBytes)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_string *retval;
+	zend_long length;
+	size_t total_size = 0;
+	size_t required_size;
+	int i;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_LONG(length)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (length < 1) {
+		zend_argument_value_error(1, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	retval = zend_string_alloc(length, 0);
+	required_size = length * sizeof(char);
+
+	while (total_size < required_size) {
+		uint64_t result;
+		size_t generated_size;
+
+		RANDOM_ENGINE_GENERATE_SIZE(randomizer->algo, randomizer->state, result, generated_size);
+
+		for (i = 0; i < generated_size; i++) {
+			ZSTR_VAL(retval)[total_size++] = (result >> (i * sizeof(char)) & 0xff);
```

Comment:
```suggestion
			ZSTR_VAL(retval)[total_size++] = (result >> (i * 8) & 0xff);
```

`sizeof(char)` is `1` by definition (thus this is a completely useless operation).

Here's a test case for the bug:

```
<?php

use Random\Engine;
use Random\Randomizer;

final class MyEngine implements Engine
{
    public function generate(): string
    {
        return "\x01\x02\x03";
    }
}

$r = new Randomizer(new MyEngine());

var_dump(\bin2hex($r->getBytes(8))); // This should output string(16) "0102030102030102"
```

Reply:
fixed 616be5b4a3a072ad7307ae0838fa31babe581578

Reply:
Late note: maybe you intended to use `CHAR_BIT`; but I guess that php-src only compiles on environments where a `char` (i.e. a byte) is always `8` bits (i.e. an "octet")